apps/src/blockly/addons/functionEditor.js

```diff
@@ -356,4 +355,41 @@ export default class FunctionEditor {
       },
     };
   }
+
+  // Copy all procedure models from the hidden definition workspace to the editor workspace,
+  // if they are not already present in the editor workspace.
+  // This is needed when the hidden definition workspace is initialized with at least one
+  // procedure, so that the editor workspace knows about those procedures.
+  setUpEditorWorkspaceProcedures() {
+    Blockly.Events.disable();
+    const procedureDefinitions = this.editorWorkspace
+      .getProcedureMap()
+      .getProcedures();
+    Blockly.getHiddenDefinitionWorkspace()
+      .getProcedureMap()
+      .getProcedures()
+      .forEach(procedure => {
+        const procedureId = procedure.getId();
+        if (
+          procedureDefinitions.filter(
+            procedureModel => procedureModel.getId() === procedureId
+          ).length === 0
```

Comment:
This might be just me, but I find this condition a little hard to wrap my head around, similar to the one in cdoUtils 😅 

Is it more readable to create a map of the procedures on the current definition and then just check if they already exist? If we do this in a couple places, we could make a little `getExistingIds` util or something to return an object containing any ids that have already been copied on a workspace?

```
const editorProcedureDefinitions = this.editorWorkspace.getProcedureMap().getProcedures();
const hiddenProcedureDefinitions = Blockly.getHiddenDefinitionWorkspace().getProcedureMap().getProcedures();

const copiedIds = {};
editorProcedureDefinitions.forEach(procedure => {
  const id = procedure.getId();
  copiedIds[id] = true;
}

hiddenProcedureDefinitions.forEach(procedure => {
  const procedureId = procedure.getId();
  if (!copiedIds) {
    const procedureModel = this.createProcedureModelForWorkspace(this.editorWorkspace, procedure);
    this.editorWorkspace.getProcedureMap().add(procedureModel);
  }
});
```

Reply:
yeah I can update this to be clearer

Reply:
lol just noticed this on the docs for [IProcedureMap](https://developers.google.com/blockly/reference/js/blockly.procedures_namespace.iproceduremap_interface.md)

> Extends: Map<string, [IProcedureModel](https://developers.google.com/blockly/reference/js/blockly.procedures_namespace.iproceduremodel_interface.md)>

so it's already a map I could have been using 🤦‍♀️ I will use that

Reply:
Unfortunately this doesn't help the `cdoUtils` case because we have to operate on the serialized data. But I am also updating that to copy over procedure models when we move procedures, which should hopefully be cleaner.

Reply:
updated this!